wordsFrequency/main.cpp

```diff
@@ -0,0 +1,129 @@
+#include <iostream>
+#include <fstream>
+#include <map>
+#include <utility>
+#include <vector>
+#include <cctype>
+#include <list>
+#include <algorithm>
+#include <limits>
+#include <optional>
+#include <set>
+#include <cassert>
+
+void addToFirstTen(std::vector<std::pair<std::string, int>>& words, const std::string& word, int frequency) {
+    bool inserted = false;
+
+    for(ulong i = 0; i < words.size(); i++) {
+        if(frequency > words[i].second)
+        {
+            auto it = words.begin() + i;
+            words.insert(it, std::pair(word, frequency));
+            inserted = true;
+            break;
+        }
+    }
+
+    if(!inserted)
+        words.push_back(std::pair(word, frequency));
+
+    if(words.size() > 10)
+        words.pop_back();
+}
+
+void collectMostFrequent(const std::string& fileName, std::vector<std::pair<std::string, int>>& firstTen) {
+    std::map<std::string, int> words;
+
+    std::ifstream file;
+    file.open (fileName);
+
+    std::string word;
+    while (file >> word)
+    {
+        if(word.length() > 3)
+            words[word]++;
+    }
+    file.close();
+
+    for (std::map<std::string, int>::iterator it = words.begin(); it != words.end(); it++ )
+        addToFirstTen(firstTen, it->first, it->second);
+
+    std::cout << "start:\n";
+    for(ulong i = 0; i < firstTen.size(); i++) {
+        std::cout << firstTen[i].first << " " << firstTen[i].second << '\n';
+    }
+}
+
+std::optional<int> getFrequency(const std::vector<std::pair<std::string, int>>& array, const std::string& word) {
```

Comment:
A me personalmente std::optional non piace granchè.
In questo caso direi che la cosa migliore da fare è tornare 0 comunque.
`std::optional` solitamente si usa per ritornare un errore o un'informazione che non può essere codificata nel range di ritorno stesso, ma questo non è il nostro caso visto che se una parola non è presente nella collezione, significa che la sua frequenza è zero appunto.